File: leetcode/Binary Tree Inorder Traversal.py
# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @param root, a tree node
    # @return a list of integers
    def inorderTraversal(self, root):
        order = []
        self.myintraversal(root, order)
        return order
    
    def myintraversal(self, root, order):
        if root:
            self.myintraversal(root.left, order)
            order.append(root.val)
            self.myintraversal(root.right, order)

File: leetcode/cpp/Binary_Tree_Inorder_Traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    //非递归解法
    //https://leetcode.com/discuss/19765/iterative-solution-in-java-simple-and-readable
    vector<int> inorderTraversal(TreeNode* root) {
        vector<int> v;
        stack<TreeNode*> stk;
        
        TreeNode *cur = root;
        while (cur != NULL || !stk.empty()) {
            while (cur) {
                stk.push(cur);
                cur = cur->left;
            }
            
            cur = stk.top();
            stk.pop();
            v.push_back(cur->val);
            cur = cur->right;
        }
        
        return v;
    }
};


//递归解法
class Solution {
public:
    vector<int> inorderTraversal(TreeNode* root) {
        vector<int> res;
        inorderTraversalHelper(root, res);
        return res;
    }
    
    void inorderTraversalHelper(TreeNode* root, vector<int> &v) {
        if (root) {
            inorderTraversalHelper(root->left, v);
            v.push_back(root->val);
            inorderTraversalHelper(root->right, v);
        }
    }
};